### ISSUE TITLE
Proposed new modeling operator (%)

### ISSUE THREAD

**Issue by perrygreenfield** [astropy/astropy#5987]
I'm proposing we add support for an additional binary operator to be used with compound models. This would be the '%' operator and would act as a argument substitution operator to an existing model. For example, supposing we have a model that takes 3 arguments, e.g., m(x, y, z) and we want to create a new model from that for which we specify one of the inputs, then:
```python
m2 = m % {1: 7.3} 
```
effectively becomes a new model that takes the value of m2(x, z) = m(x, 7.3, z).

LIkewise, we could have
```python
m3 = m % {1: 7.3, 2: 3.2} # m3(x) = m(x, 7.3, 3.2)
```
Essentially, this is a "currying"-like operator. 

Why is this useful?

This allows more restricted versions of more general models to be used conveniently. It has explicitly useful application with WCS models where one may define a complex model that has special extra coordinates such as spectral order, position in the slit, etc. Then when one wants to define a specific  WCS transform for specific spectral order, one do away entirely with requiring that value having to be supplied in the transformation. In this way there can be one master transformation in all possible generality, and then one can define more specific instances of transformations without having to build an entirely new model. Simply base a model definition on the general model in a more restricted way.

**Comment by Cadair** [comment]
This seems conceptually very similar to partial functions, i.e `functools.partial`. I am assuming there are many edge cases for modelling where actually using `functools.partial` would not work, but couldn't something like `partial_model` be implemented instead of another binary operator?

**Comment by perrygreenfield** [comment]
Sure, one could use a similar function, but since compound models heavily use operators, an operator approach seems most natural. The actual functools.partial probably wouldn't work as is, and it presumes that the arguments you want to set are first in order (an unnecessary restriction). I'd find the above approach more general and likely easier to understand. Finally, implementing support for an operator in the existing (or the alternate) compound operator is pretty simple.

**Comment by MSeifert04** [comment]
I dislike inventing use cases for operators and neither modulo operation or string-formatting seem to match this use-case. Also in most cases a method more clearly describes the action.

It's annoying that `functools.partial` does not support missing positional arguments (that's why I recently build one myself) but you can always provide them as keyword-arguments.

**Comment by perrygreenfield** [comment]
I'm pretty sure that functools.partial wouldn't come close to working (there is associated info of a compound object that needs to be made consistent with the change, and that tool can't possibly know about that). But it is pretty close to string formatting in the general idea. It's use in string formatting is effectively substitution, and that's what is basically being done here. I don't think any other operator come close to meaning that.

**Comment by nden** [comment]
This functionality was implemented as a function `fix_inputs` in #9135
